client/asset/dcr/dcr.go

```diff
@@ -850,32 +873,42 @@ func (dcr *ExchangeWallet) Balance() (*asset.Balance, error) {
 
 // FeeRate satisfies asset.FeeRater.
 func (dcr *ExchangeWallet) FeeRate() uint64 {
-	if dcr.wallet.SpvMode() {
-		return 0 // EstimateSmartFeeRate needs dcrd passthrough
-	}
-	// Requesting a rate for 1 confirmation can return unreasonably high rates.
-	rate, err := dcr.feeRate(2)
+	const confTarget = 2
+	rate, err := dcr.feeRate(confTarget)
 	if err != nil {
-		dcr.log.Errorf("Failed to get fee rate: %v", err)
-		return 0
+		dcr.log.Errorf("feeRate error: %v", err)
 	}
 	return rate
 }
 
 // FeeRate returns the current optimal fee rate in atoms / byte.
 func (dcr *ExchangeWallet) feeRate(confTarget uint64) (uint64, error) {
-	feeEstimator, is := dcr.wallet.(FeeRateEstimator)
-	if !is {
-		return 0, fmt.Errorf("fee rate estimation unavailable")
-	}
 	// estimatesmartfee 1 returns extremely high rates on DCR.
 	if confTarget < 2 {
 		confTarget = 2
 	}
-	estimatedFeeRate, err := feeEstimator.EstimateSmartFeeRate(dcr.ctx, int64(confTarget), chainjson.EstimateSmartFeeConservative)
+	if feeEstimator, is := dcr.wallet.(FeeRateEstimator); is {
+		ratePerKB, err := feeEstimator.EstimateSmartFeeRate(dcr.ctx, int64(confTarget), chainjson.EstimateSmartFeeConservative)
+		if err == nil {
+			return convertFeeToUint(ratePerKB)
+		} else {
+			dcr.log.Errorf("Failed to get fee rate with estimate smart fee rate: %v", err)
+		}
+	}
+	// Either SPV wallet or EstimateSmartFeeRate failed.
+	if !dcr.apiFeeFallback {
+		return 0, fmt.Errorf("fee rate estimation unavailable")
+	}
+	dcr.log.Debug("Retrieving fee rate from external API")
+	ratePerKB, err := externalFeeEstimator(dcr.ctx, dcr.network, confTarget)
```

Comment:
I'm just realizing that we're not doing any sanity checking or thresholding on the result from the external source.  They could be malicious or maybe they changed the API from DCR/kB -> atoms/kB (or whatever) without warning.
The external source might also say 0.

I think after calling `externalFeeEstimator` we need to both catch zero and filter the result of `convertFeeToUint` through `feeRateWithFallback`

Reply:
Yeah, you are right, it was returning 1, which would not go into `feeRateWithFallback`. Returning 0 now if it is 0.

I am not checking for a upper limit though, should I?

Reply:
sounds reasonable. Checking if it is greater than  `dcr.feeRateLimit` now